Modern/ops/src/main/java/org/bonej/ops/ellipsoid/FindEllipsoidFromBoundaryPoints.java

```diff
@@ -0,0 +1,343 @@
+        axisContribution.m21(normalizedAxis.y()*normalizedAxis.z()/(radius*radius));
+
+        return axisContribution;
+    }
+
+
+    private static ValuePair<VertexWithNormal,Double> calculateQ(final List<Vector3d> candidateQs, final VertexWithNormal p) {
+        List<ValuePair<VertexWithNormal, Double>> candidateQAndRs = candidateQs.stream().map(q -> calculatePossibleQAndR(q,p)).filter(qnr -> qnr.getB()>0).collect(Collectors.toList());
+
+        candidateQAndRs.sort(Comparator.comparingDouble(ValuePair::getB));
+
+        if(candidateQAndRs.size() > 0)
+            return candidateQAndRs.get(0);
+        else
+            return null;
+    }
+
+    private static double euclideanDistance(Vector3d q, Vector3d p) {
+        Vector3d distance = new Vector3d(q);
+        distance.sub(p);
+        return distance.length();
+    }
+
+    private static ValuePair<VertexWithNormal,Double> calculatePossibleQAndR(Vector3d x, VertexWithNormal p)
+    {
+        Vector3d xMinusP = new Vector3d(p.getVertex());
+        xMinusP.mul(-1);
+        xMinusP.add(x);
+        double distanceSquared = xMinusP.lengthSquared();
+        double scalarProduct = xMinusP.dot(p.getNormal());
+
+        if(scalarProduct<=0.0) return new ValuePair<>(null,-1.0);
+
+        double radius = distanceSquared/(2*scalarProduct);
+
+        Vector3d centre = new Vector3d(p.getNormal());
+        centre.mul(radius);
+        centre.add(p.getVertex());
+
+        Vector3d cMinusX = new Vector3d(x);
+        cMinusX.mul(-1);
+        cMinusX.add(centre);
+        return new ValuePair<>(new VertexWithNormal(new ValuePair<>(x,cMinusX)),radius);
+    }
+
+    private static ValuePair<Vector3d, Double> calculateSurfacePointAndGreekCoefficient(final Matrix4d Q1, final Matrix4d Q2, final List<Vector3d> vertices)
+    {
+        final List<ValuePair<Vector3d,Double>> candidatePointsAndCoefficients = vertices.stream().map(v -> calculateCandidateSurfacePointAndGreekCoefficient(Q1, Q2, v)).filter(a -> !a.getB().isNaN()).collect(Collectors.toList());
+        candidatePointsAndCoefficients.sort(Comparator.comparingDouble(ValuePair::getB));
+        if(candidatePointsAndCoefficients.size() > 0)
+            return candidatePointsAndCoefficients.get(0);
+        else
+            return null;
+    }
+
+    private static double calculateXtQX(final Matrix4d Q2, Vector3d x)
+    {
+        Vector4d x4d = new Vector4d(x,1);
+        Vector4d Q2X = new Vector4d(x4d);
+        Q2.transform(Q2X);
+        return x4d.dot(Q2X);
+    }
+
+    private static ValuePair<Vector3d, Double> calculateCandidateSurfacePointAndGreekCoefficient(final Matrix4d Q1, final Matrix4d Q2, final Vector3d x)
+    {
+        double xtQ2X = calculateXtQX(Q2, x);
+        if(xtQ2X>=0) return new ValuePair<>(new Vector3d(), Double.NaN);
+
+        double xtQ1X = calculateXtQX(Q1, x);
+        return new ValuePair<>(new Vector3d(x),-xtQ1X/xtQ2X);
+    }
+
+
+}
+
+class VertexWithNormal {
```

Comment:
I'm surprised that having a second class definition in the same file hasn't caused problems. I don't see much value to this rewrapping of `ValuePair<Vector3d, Vector3d>` especially if you rewrite your op to use ``ValuePair<Vector3dc, Vector3dc>`` like I suggested earlier.

If you still want to have the convenience of calling `getVertex()` and `getNormal()` I suggest replacing them with simple `Functions`:
```
private static final Function<ValuePair<Vector3dc, Vector3dc>, Vector3d> GET_NORMAL = pair -> new Vector3d(pair.b);
private static final Function<ValuePair<Vector3dc, Vector3dc>, Vector3d> GET_VERTEX = pair -> new Vector3d(pair.a);
```
The setter is only used once, so I don't think it's necessary.